pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -104,125 +110,117 @@ func genObservationID() (uint32, error) {
+			return expInput, err
 		}
 		// Currently, supporting only IPv4 for Flow Aggregator.
 		ip := hostIPs[0].To4()
 		if ip != nil {
-			// Update the collector address with resolved IP of flow aggregator
 			collectorAddr = net.JoinHostPort(ip.String(), defaultIPFIXPort)
 		} else {
-			return fmt.Errorf("resolved Flow Aggregator address %v is not supported", hostIPs[0])
+			return expInput, fmt.Errorf("resolved Flow Aggregator address %v is not supported", hostIPs[0])
 		}
 	}
 
 	// TODO: This code can be further simplified by changing the go-ipfix API to accept
 	// collectorAddr and collectorProto instead of net.Addr input.
-	var expInput exporter.ExporterInput
 	if collectorProto == "tcp" {
-		collector, err := net.ResolveTCPAddr("tcp", collectorAddr)
+		expInput.CollectorAddr, err = net.ResolveTCPAddr("tcp", collectorAddr)
 		if err != nil {
-			return err
+			return expInput, err
 		}
 		// TCP transport does not need any tempRefTimeout, so sending 0.
 		// tempRefTimeout is the template refresh timeout, which specifies how often
 		// the exporting process should send the template again.
-		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
-			ObservationDomainID: obsID,
-			TempRefTimeout:      0,
-			PathMTU:             0,
-			IsEncrypted:         false,
-		}
+		expInput.TempRefTimeout = 0
 	} else {
-		collector, err := net.ResolveUDPAddr("udp", collectorAddr)
+		expInput.CollectorAddr, err = net.ResolveUDPAddr("udp", collectorAddr)
 		if err != nil {
-			return err
+			return expInput, err
 		}
 		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
-		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
-			ObservationDomainID: obsID,
-			TempRefTimeout:      1800,
-			PathMTU:             0,
-			IsEncrypted:         false,
+		expInput.TempRefTimeout = 1800
+	}
+	expInput.PathMTU = 0
+	expInput.IsEncrypted = false
+
+	return expInput, nil
+}
+
+func NewFlowExporter(connStore *connections.ConnectionStore, records *flowrecords.FlowRecords, collectorAddr string, collectorProto string, activeFlowTimeout time.Duration, inactiveFlowTimeout time.Duration, v4Enabled bool, v6Enabled bool) (*flowExporter, error) {
+	// Initialize IPFIX registry
+	registry := ipfix.NewIPFIXRegistry()
+	registry.LoadRegistry()
+
+	// Prepare input args for IPFIX exporting process.
+	expInput, err := prepareExporterInputArgs(collectorAddr, collectorProto)
+	if err != nil {
+		return nil, err
+	}
+
+	return &flowExporter{
+		connStore:           connStore,
+		flowRecords:         records,
+		registry:            registry,
+		v4Enabled:           v4Enabled,
+		v6Enabled:           v6Enabled,
+		exporterInput:       expInput,
+		activeFlowTimeout:   activeFlowTimeout,
+		inactiveFlowTimeout: inactiveFlowTimeout,
+	}, nil
+}
+
+// Run exports flow records at a given flow export frequency.
+func (exp *flowExporter) Run(stopCh <-chan struct{}) {
+	go wait.Until(exp.Export, time.Second, stopCh)
+
+	<-stopCh
+}
```

Comment:
Yes, will modify the comment.

As flows can start at any time, they will be timed out at different intervals. If we do not do the scanning every second, we may miss exact timeouts for some flows and they will be timed out at higher time intervals. 
Do you have any better alternative in mind?

Reply:
There are other implementations we can consider, like having a priority queue ordered based on the last export timestamp. But the most important thing at that stage is to have a unit tests to benchmark this. I don't remember the details of the flow exporter implementation, but if the number of flow records is on par with the number of connections tracked, then a scale of 512K - 1M is not unreasonable (https://www.projectcalico.org/when-linux-conntrack-is-no-longer-your-friend). That applies to the rest of the FlowExporter code as well. Now maybe we are far from having done all the possible optimizations. For example, are we doing anything special for TCP connections in the `TIME_WAIT` state, which can account for a large fraction of tracked connections? It seems that at some point, we don't need to include them in our connection map any more.

Reply:
Priority queue is a good suggestion. It can be explored to see if a one second scan can be avoided.

Yes, benchmarking flow exporter using uint tests with test connections in connection map or test flow records in record map is a good first step. As flows are very diverse (different proto, different status, different state etc.); generating these test connections could be tricky. Any thoughts here? 

Yes, good point for connections with `TIME_WAIT` state for optimization. We should export the flow record and delete the connection in connection map. Furthermore, we should not add any connections with `TIME_WAIT` state.